### PR TITLE
fix(trans): align render callback and 'component' prop

### PR DESCRIPTION
**Pull request by vonovak**
# Description

The `Trans` component accepts `component` and `render` props that have pretty much the same purpose.

However, when you use the `render` callback, the callback receives 

```
export type TransRenderProps = {
   id: string
   translation: React.ReactNode
   message?: string | null
   isTranslated: boolean
 }
```

and when you use `component`, it only receives `{ children: React.ReactNode }` (where `children = translation`).

To me, it makes sense that these two approaches are unified, and you receive the same information when you use `render` and `component` props.

This PR does just that, without introducing another TS type, which means that the `translation` and `children` are duplicated in the `i18nprops: TransRenderProps` object. I don't think it's a huge deal but we could also introduce separate types for `render` and `component`.

I also removed a few `as` casts to improve type safety

Let me know what you think and I can update the docs as well, as they are not very detailed in describing these props.

## Types of changes

[//]: # (What types of changes does your code introduce to Lingui?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

Fixes # (issue)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/lingui/js-lingui/blob/main/CONTRIBUTING.md) and [CODE_OF_CONDUCT](https://github.com/lingui/js-lingui/blob/main/CODE_OF_CONDUCT.md) docs
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: packages/react/src/Trans.test.tsx [hunk 1]
```diff
@@ -54,13 +54,27 @@ describe("Trans component", function () {
...
   })
 
-  it("should throw a console.error if using twice props", function () {
+  it("should log a console.error if using `render` and `component` props at the same time", function () {
+    const RenderChildrenInSpan = ({ children }: TransRenderProps) => (
+      <span>{children}</span>
+    )
+
     const originalConsole = console.error
     console.error = jest.fn()
```
Comment (thekip) [inline packages/react/src/Trans.test.tsx:63]: there is a special function `mockConsole` is `jest-mocks` package in monorepo. You can use it to make code more unified.

File: packages/react/src/Trans.tsx [hunk 2]
```diff
@@ -18,9 +21,10 @@ export type TransProps = {
   components?: { [key: string]: React.ElementType | any }
   formats?: Record<string, unknown>
   children?: React.ReactNode
+} & MaximumOneOf<{
   component?: React.ComponentType<TransRenderProps>
   render?: (props: TransRenderProps) => React.ReactElement<any, any> | null
-}
+}>
```
Comment (thekip) [inline packages/react/src/Trans.tsx:27]: interesting, more simple form will not work? 

```ts
export type TransProps = { [...] } 
   & (
       {component?: React.ComponentType<TransRenderProps>} 
       |  { render?: (props: TransRenderProps) => React.ReactElement<any, any> | null}
    )
```

File: packages/react/src/Trans.test.tsx [hunk 3]
```diff
@@ -54,13 +54,27 @@ describe("Trans component", function () {
     console.error = originalConsole
   })
 
-  it("should throw a console.error if using twice props", function () {
+  it("should log a console.error if using `render` and `component` props at the same time", function () {
```
Comment (thekip) [inline packages/react/src/Trans.test.tsx:57]: could you also change all `function () {` to `() => {` in the test file.

**Review by thekip — CHANGES_REQUESTED** [review]

**Comment by thekip** [comment]
The changes itself is looks good to me. As i mentioned earlier, i'm not too much in the React itself, so i could not evaluate a value of this PR from react users point of view. But from a code perspective, all looks good, except what i've written.

File: packages/react/src/Trans.tsx (hunk 2 above)
Comment (vonovak) [reply packages/react/src/Trans.tsx:27]: tried that first, and it didn't work as well

File: packages/react/src/Trans.test.tsx [hunk 4]
```diff
@@ -54,17 +54,29 @@ describe("Trans component", function () {
...
+  it("should log a console.error if using `render` and `component` props at the same time", () => {
+    const RenderChildrenInSpan = ({ children }: TransRenderProps) => (
+      <span>{children}</span>
+    )
 
-    // @ts-expect-error testing the error
-    renderWithI18n(<Trans render="div" component="span" id="Some text" />)
-    expect(console.error).toHaveBeenCalled()
-    console.error = originalConsole
+    const consoleErrorSpy = jest.spyOn(console, "error")
```
Comment (thekip) [inline packages/react/src/Trans.test.tsx]: will this print error in the console while testing? If yes - it's not good. These errors, very annoying when you run full suite

**Comment by vonovak** [comment]
@thekip addressed all comments and ended up rebasing on top of next

**Comment by andrii-bodnar** [comment]
@thekip could you please check if the updates after comments are ok for you? I'd like to include it in `v4.0.0-next.4`

**Review by andrii-bodnar — APPROVED** [review]
@vonovak thank you!

> I can update the docs as well, as they are not very detailed in describing these props.

Are you going to update it within the current PR or in a separate one?

**Comment by vonovak** [comment]
> @vonovak thank you!
> 
> > I can update the docs as well, as they are not very detailed in describing these props.
> 
> Are you going to update it within the current PR or in a separate one?

A separate one, thank you

**Pull request merged**